pkg/kubelet/pluginmanager/operationexecutor/operation_generator.go

```diff
@@ -135,12 +144,33 @@ func (og *operationGenerator) GenerateRegisterPluginFunc(
 func (og *operationGenerator) GenerateUnregisterPluginFunc(
 	socketPath string,
 	pluginHandlers map[string]cache.PluginHandler,
+	pathToHandlers *cache.SocketPluginHandlers,
 	actualStateOfWorldUpdater ActualStateOfWorldUpdater) func() error {
 
 	unregisterPluginFunc := func() error {
 		client, conn, err := dial(socketPath, dialTimeoutDuration)
 		if err != nil {
-			return fmt.Errorf("UnregisterPlugin error -- dial failed at socket %s, err: %v", socketPath, err)
+			var handlerWithName cache.NamedPluginHandler
+			handlerFound := false
+			pathToHandlers.Lock()
+			if len(pathToHandlers.Handlers) != 0 {
+				handlerWithName, handlerFound = pathToHandlers.Handlers[socketPath]
+			}
+			pathToHandlers.Unlock()
+
+			if !handlerFound {
+				return fmt.Errorf("unable to DeRegister - there is no handler for %s", socketPath)
+			}
+			// We remove the plugin to the actual state of world cache before calling a plugin consumer's Unregister handle
+			// so that if we receive a register event during Register Plugin, we can process it as a Register call.
+			actualStateOfWorldUpdater.RemovePlugin(socketPath)
+
+			handlerWithName.Handler.DeRegisterPlugin(handlerWithName.Name)
+			pathToHandlers.Lock()
+			delete(pathToHandlers.Handlers, socketPath)
+			pathToHandlers.Unlock()
+			klog.V(4).Infof("DeRegisterPlugin called for %s on %v", handlerWithName.Name, handlerWithName.Handler)
+			return nil
```

Comment:
I think we can merge this part with line 192 to 194 by moving this outside of the if statement